### PR TITLE
Add overlay for deployment

### LINKED ISSUE #14: Expose resources for awx-controller-manager in chart
Migrating from https://github.com/ansible/awx-operator/issues/1919

Summary

As noted in https://github.com/ansible/awx-operator/issues/1115, GKE Autopilot automatically changes pod resource requests and/or limits. The specific defect described in that issue occurs because Autopilot clusters with the `bursting` feature `disabled` automatically (and transparently) change pod `resources` to set `limits` equal to the `requests` value when both values are defined. This means that `awx-operator-controller-manager` is left with insufficient memory, and will always be `OOMKilled`.

In my case, I deploy the `awx-operator` helm chart with an Ansible playbook, and thus I can use a `kubernetes.core.k8s_json_patch` task on the `Deployment` to set the `requests` value equal to the `limits` value immediately after the helm chart is deployed, effectively working around the problem.  However, this does introduce the issue that any change that redeploys the chart will also have to re-patch the Deployment.

What https://github.com/ansible/awx-operator/issues/1115 didn't spell out, and I'm specifically requesting here, is the ability to adjust the `resources` `requests` and `limits` for the `awx-manager` pod in the `Deployment` from the helm chart values directly. This would allow us to handle this issue before deploying the chart, and prevent the play from repeatedly adjusting these values in subsequent runs.

Basically, as D1StrX notes there, the CRD AWX should include a section for `awx_manager_resource_requirements`.

### PR DESCRIPTION
**Pull request by miles-w-3**
Closes #14. Deployment resource is now fully templatable via .Values.controller.spec, .labels, and .annotations. 

Only regression I see is that the previously spliced-in template for .Values.Operator.replicas will not work, but I believe this setup is more scalable

### CODE REVIEW

File: .helm/starter/templates/controller/controller-overlay.yaml
Comment (oraNod) [inline .helm/starter/templates/controller/controller-overlay.yaml:15]: Sorry to be a stickler but could you please add an empty line to the EOF?

**Review by oraNod — COMMENTED** [review]
Thanks for the PR @miles-w-3 I've left a super minor nit for now. This seems fine to me but I'm at a conf for the next few days and would prefer to take a closer look when I return. I won't leave it hanging for too long though. Cheers.

File: .helm/starter/templates/controller/controller-overlay.yaml
Comment (miles-w-3) [reply .helm/starter/templates/controller/controller-overlay.yaml]: @oraNod done

**Review by oraNod — APPROVED** [review]
I'm not that familiar with the use case here but I think these changes seem fine. Thanks for the PR @miles-w-3

**Comment by miles-w-3** [comment]
fwiw, the use case is if you need to override Kubernetes specs like the memory limit, image tag, environment variables, etc. Under the current system with the operator SDK, it was designed that you change/add those fields with kustomize, but it's not really compatible with the helm packaging solution.

As I remember, there were a lot of issues back when this image was in the main repo that were related to "why can't I change these deployment fields with the helm chart"

**Comment by miles-w-3** [comment]
Added commit containing documentation and a check for the `python3` variable in path

**Comment by ChrisLee-JackHenry** [comment]
Is this waiting for something specific?

**Comment by oraNod** [comment]
Closed and re-opened to kick off ci with the changes from https://github.com/ansible-community/awx-operator-helm/pull/25

@ChrisLee-JackHenry sorry I missed the comment and didn't respond earlier. this has been on hold until we fixed ci.

**Comment by oraNod** [comment]
@miles-w-3 Hey, looks like there's a type error in here:

```
Error:  templates/: template: awx-operator/templates/controller/controller-overlay.yaml:4:53: executing "awx-operator/templates/controller/controller-overlay.yaml" at <$.Values.controller.labels>: wrong type for value; expected map[string]interface {}; got interface {}
```

File: .helm/starter/values.yaml
Comment (oraNod) [inline .helm/starter/values.yaml:21]: Sorry to be nit-picky but could you add an extra line at the end of the file here please?

**Comment by miles-w-3** [comment]
thanks for the review, will rebase and address these fixes.

**Comment by miles-w-3** [comment]
@oraNod addressed the error and standardized the behavior to merge at the root of the doc as I had initially written in the readme

File: Makefile
Comment (rooftopcellist) [inline Makefile:163]: @miles-w-3 I think this is a good change, but it may be better to call this parameter something other than `controller` because it may cause confusion with the downstream of awx (which is called Automation Controller and referred to as controller).  

I know this is referring to the k8s controller-manager (how k8s docs refer to the operator deployment)... But in our unique position, maybe we could call it `operator-controller`, or just `operator`? , or even `manager`, as that is what the deployment yaml file is called (config/manager/manager.yaml).  

What do you think?

Comment (miles-w-3) [reply Makefile:163]: I defer to your judgement, operator-controller or manager-deployment make sense to me and I can make the change. Can also update the template names in the chart to avoid confusion

Comment (miles-w-3) [reply Makefile:163]: @rooftopcellist I have updated the pr changing the key to `operator-controller`

**Review by oraNod — APPROVED** [review]
Thanks @miles-w-3

File: README.md
Comment (schen1) [inline README.md:253]: While the example of the replica is valid. This is not really a common scenario to customize the operator replica (maybe to scale up/scale down). One more common use case is to specify a different container image for each container as seen in [this issue](https://github.com/ansible-community/awx-operator-helm/issues/17) or [that one](https://github.com/ansible-community/awx-operator-helm/issues/27) to cover disconnected/restricted environments.

What do you think? Could you maybe provide an example on how to do that @miles-w-3 

Thanks for the PR btw!

File: Makefile
Comment (oraNod) [reply Makefile:163]: @miles-w-3 Thanks for making the change. That looks good.

File: README.md
Comment (oraNod) [reply README.md:253]: @schen1 Perhaps we should capture these items in a follow up to improve the docs. While I think extensive documentation is always a good thing, I'm not sure we should hold up this PR much longer. Any thoughts @miles-w-3 ?

**Review by oraNod — APPROVED** [review]
Looks good. Thanks for the work on this @miles-w-3

**Review by schen1 — APPROVED** [review]

File: README.md
Comment (schen1) [reply README.md:253]: fine for me :)

**Comment by oraNod** [comment]
Follow up issue for docs: https://github.com/ansible-community/awx-operator-helm/issues/29

File: README.md
Comment (miles-w-3) [reply README.md:253]: I am fine either way. The use case for overriding an image tag is tricky because of the way lists are merged in helm. 

Lists are overridden instead of merged for each layer. So under this proposal, if you want to change the image version, you have to override the entire container list spec. For that reason, this solution is more optimal for adding things like annotations or service account specs to the controller's pods. 

If there is desire for more targeted overrides of specific containers, that could be another feature for a template that iterates through a list, matches a container name, and then merges a set override

Comment (schen1) [reply README.md:253]: Thanks for the details and the great contribution. Then it makes sense to have it as a different feature/PR

**Pull request merged**
